cloud/services/publicips/publicips.go

```diff
@@ -18,61 +18,124 @@ package publicips
-			ctx,
-			s.Scope.ResourceGroup(),
-			ip.Name,
-			network.PublicIPAddress{
-				Sku:      &network.PublicIPAddressSku{Name: network.PublicIPAddressSkuNameStandard},
-				Name:     to.StringPtr(ip.Name),
-				Location: to.StringPtr(s.Scope.Location()),
-				PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
-					PublicIPAddressVersion:   network.IPv4,
-					PublicIPAllocationMethod: network.Static,
-					DNSSettings: &network.PublicIPAddressDNSSettings{
-						DomainNameLabel: to.StringPtr(strings.ToLower(ip.Name)),
-						Fqdn:            to.StringPtr(ip.DNSName),
-					},
+		ip, err := s.Client.Get(ctx, s.Scope.ResourceGroup(), publicIPSpec.Name)
+		if err != nil && !ip.IsHTTPStatus(http.StatusNotFound) {
+			return errors.Wrapf(err, "failed to fetch public ip '%s' to check if we should create", publicIPSpec.Name)
+		}
+
+		if ip.IsHTTPStatus(http.StatusOK) {
+			klog.V(2).Infof("found existing public ip %s, returning early", publicIPSpec.Name)
+			return nil
+		}
+
+		ipSpec := network.PublicIPAddress{
+			Sku:      &network.PublicIPAddressSku{Name: network.PublicIPAddressSkuNameStandard},
+			Name:     to.StringPtr(publicIPSpec.Name),
+			Location: to.StringPtr(s.Scope.Location()),
+			PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
+				PublicIPAddressVersion:   network.IPv4,
+				PublicIPAllocationMethod: network.Static,
+				DNSSettings: &network.PublicIPAddressDNSSettings{
+					DomainNameLabel: to.StringPtr(strings.ToLower(publicIPSpec.Name)),
+					Fqdn:            to.StringPtr(publicIPSpec.DNSName),
 				},
 			},
-		)
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(s.Scope.ResourceGroup()),
+				Role:        to.StringPtr(infrav1.CommonRole),
+				Additional:  s.Scope.AdditionalTags(),
+			})),
+		}
 
-		if err != nil {
-			return errors.Wrap(err, "cannot create public IP")
+		// https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-standard-availability-zones#zone-redundant-by-default
+
+		if err := s.Client.CreateOrUpdate(
+			ctx,
+			s.Scope.ResourceGroup(),
+			publicIPSpec.Name,
+			ipSpec,
+		); err != nil {
+			return errors.Wrap(err, "cannot create public ip")
 		}
-		klog.V(2).Infof("successfully created public IP %s", ip.Name)
+
+		klog.V(2).Infof("successfully created public IP %s", publicIPSpec.Name)
 	}
 	return nil
 }
 
 // Delete deletes the public IP with the provided scope.
 func (s *Service) Delete(ctx context.Context) error {
-	for _, ip := range s.Scope.PublicIPSpecs() {
-		klog.V(2).Infof("deleting public IP %s", ip.Name)
-		err := s.Client.Delete(ctx, s.Scope.ResourceGroup(), ip.Name)
+	for _, publicIPSpec := range s.Scope.PublicIPSpecs() {
+		klog.V(2).Infof("deleting public IP %s", publicIPSpec.Name)
+		err := s.Client.Delete(ctx, s.Scope.ResourceGroup(), publicIPSpec.Name)
```

Comment:
rebase err? seems like we're deleting the IP twice in this block

Reply:
whoops